src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -611,15 +558,49 @@ private static string UrlPathEncodeImpl(string value)
                 return value;
             }
 
-            // recurse in case there is a query string
-            int i = value.IndexOf('?');
-            if (i >= 0)
+            int i = value.AsSpan().IndexOfAnyExceptInRange((char)0x21, (char)0x7F);
+            if (i < 0)
+            {
+                return value;
+            }
+
+            int indexOfQuery = value.IndexOf('?');
+            if (indexOfQuery >= 0 && indexOfQuery < i)
```

Comment:
```suggestion
            if ((uint)indexOfQuery < (uint)i)
```
`i` is a positive number, according the check in L562.
The `uint` cast of `indexOfQuery` will "move" a negative number in the range > int.MaxValue, thus you can save a comparison by this cast-trick.